src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs

```diff
@@ -2034,6 +2034,694 @@ static bool Set(out int x)
+        _ = c1?.M(x = 0) ?? throw new System.Exception();
+        x.ToString();
+    }
+
+    bool M(object obj) { return true; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_Cast()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        _ = (object)c1?.M(x = 0) ?? throw new System.Exception();
+        x.ToString();
+    }
+
+    C M(object obj) { return this; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_ImplicitReferenceConv()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        object obj = c1?.M(x = 0) ?? (object)(x = 0);
+        x.ToString();
+    }
+
+    C M(object obj) { return this; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_BoxingConv()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1)
+    {
+        int x;
+        object obj = c1?.M(x = 0) ?? c1.M(x = 0);
+        x.ToString();
+    }
+
+    bool? M(object obj) { return null; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_UserDefinedConv_01()
+        {
+            var source = @"
+struct S { }
+
+struct C
+{
+    public static implicit operator S(C? c)
+    {
+        return default(S);
+    }
+
+    void M1(C? c1)
+    {
+        int x;
+        S s = c1?.M2(x = 0) ?? c1.Value.M3(x = 0);
+        x.ToString(); // 1
+    }
+
+    C M2(object obj) { return this; }
+    S M3(object obj) { return this; }
+}
+";
+            // PROTOTYPE(improved-definite-assignment):
+            // Arguably, in this scenario, the RHS of the `??` should be visited in an unreachable state
+            CreateCompilation(source).VerifyDiagnostics(
+                // (15,9): error CS0165: Use of unassigned local variable 'x'
+                //         x.ToString(); // 1
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(15, 9));
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_UserDefinedConv_02()
+        {
+            var source = @"
+class B { }
+class C
+{
+    public static implicit operator B(C c) => new B();
+
+    void M1(C c1)
+    {
+        int x;
+        B b = c1?.M1(x = 0) ?? c1!.M2(x = 0);
+        x.ToString(); // 1
+    }
+
+    C M1(object obj) { return this; }
+    B M2(object obj) { return new B(); }
+}
+";
+            // If the LHS of a `??` is subject to a user-defined conversion whose parameter is not a non-nullable value type
+            // we can't propagate out the "state when not null"
+            // because we can't know whether the conditional access itself was non-null.
```

Comment:
This should be valid. We can and do know whether the conditional access was non-null.

Reply:
I do actually wonder whether we correctly handle the case of the implicit operator converting to null even though the underlying expression wasn't... 🤔

---
In reply to: [597328758](https://github.com/dotnet/roslyn/pull/51567#discussion_r597328758) [](ancestors = 597328758)

Reply:
The answer is no, we do not. Filed [https://github.com/dotnet/roslyn/issues/51975](https://github.com/dotnet/roslyn/issues/51975).

---
In reply to: [597329999](https://github.com/dotnet/roslyn/pull/51567#discussion_r597329999) [](ancestors = 597329999,597328758)

Reply:
> This should be valid. We can and do know whether the conditional access was non-null.

I'm a little confused by this assertion. Does the null test occur before the implicit conversion?

Reply:
Ah, the linked issue says this. Thanks, that clears up a lot, I'll edit the spec/implementation/tests tomorrow to reflect this.